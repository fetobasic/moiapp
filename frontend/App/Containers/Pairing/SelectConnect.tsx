import React, { useCallback } from 'react';
import { ScrollView, View, Text, StyleSheet } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { NativeStackScreenProps } from '@react-navigation/native-stack';

import { Button, HeaderSimple as Header, InfoRow } from 'App/Components';

import { ApplicationStyles, Colors, Fonts, Metrics } from 'App/Themes';
import { PairingStackParamList } from 'App/Types/NavigationStackParamList';
import renderElement from 'App/Services/RenderElement';

import DirectConnectIcon from 'App/Images/Icons/directConnect.svg';
import CloudConnectIcon from 'App/Images/Icons/cloudConnect.svg';
import { useAppSelector, useMount, usePairingNavigation } from 'App/Hooks';
import { ConnectionType } from 'App/Types/ConnectionType';
import { showConfirm } from 'App/Services/Alert';

type Props = NativeStackScreenProps<PairingStackParamList, 'SelectConnect'>;

function SelectConnect({ route }: Props) {
  const navigation = usePairingNavigation('SelectConnect');
  const { user } = useAppSelector((state) => ({
    user: state.auth.user,
  }));

  const isCloudEnabled = !!user?.email;

  const navigate = useCallback(
    (connectionType: ConnectionType) => {
      if (!isCloudEnabled && connectionType === 'cloud') {
        showConfirm(
          'Log In',
          'You need to Log In with Goal Zero account to cloud connect.',
          () =>
            navigation.navigate('LoginNav', {
              ...route.params,
              connectionType: 'cloud',
              onSuccess: (params) => {
                navigation.navigate('SelectWifiNetwork', { ...route.params, ...params });
              },
            }),
          { confirmTitle: 'Continue' },
          undefined,
          true,
        );
        return;
      }

      const page =
        route.params?.dataTransferType === 'bluetooth'
          ? connectionType === 'cloud'
            ? 'SelectWifiNetwork'
            : 'ConnectYeti'
          : 'FindWifiNetwork';

      navigation.navigate(page, { ...route.params, connectionType });
    },
    [isCloudEnabled, navigation, route.params],
  );

  useMount(() => {
    if (route.params?.connectionType) {
      navigate(route.params.connectionType);
    }
  });

  return (
    <SafeAreaView edges={['bottom']} style={ApplicationStyles.mainContainer}>
      <Header title="Select Connection" onBackPress={() => navigation.navigate('AddNewDevice')} />
      <View style={ApplicationStyles.container}>
        <View style={ApplicationStyles.flex}>
          <ScrollView showsVerticalScrollIndicator={false}>
            <InfoRow
              withBorder={false}
              icon={renderElement(<CloudConnectIcon color={Colors.transparentWhite('0.6')} />)}
              title="Cloud Connect"
              subTitle={renderElement(
                <>
                  <Text style={styles.textSubTitle}>Best for home use</Text>
                  <Text style={styles.textSubTitle}>
                    Use Cloud Connect to set up your device through your internet network. This lets you control your
                    device from anywhere you have internet access, using your phone. It also allows feature updates.
                  </Text>
                </>,
              )}
              body={renderElement(
                <Text style={styles.textSubTitle}>
                  <Text style={{ color: Colors.portWarning }}>Note: </Text>Cloud Connect requires an internet connection
                  and Goal Zero log in.
                </Text>,
              )}
            />
            <InfoRow
              withBorder={false}
              style={styles.sectionTop}
              icon={renderElement(<DirectConnectIcon color={Colors.transparentWhite('0.6')} />)}
              title="Direct Connect"
              subTitle={renderElement(
                <>
                  <Text style={styles.textSubTitle}>Best for off-grid use</Text>
                  <Text style={styles.textSubTitle}>
                    Use Direct Connect to set up your device through Bluetooth. You will be able to control your device
                    from your phone when you are within 50 feet of the device.
                  </Text>
                </>,
              )}
              body={renderElement(
                <Text style={styles.textSubTitle}>
                  <Text style={{ color: Colors.portWarning }}>Note: </Text>Automatic feature updates are not available
                  through Direct Connect.
                </Text>,
              )}
            />
          </ScrollView>
        </View>
        <View style={styles.sectionButtons}>
          <Button testID="btnDirect" style={styles.btnLeft} inverse title="Direct" onPress={() => navigate('direct')} />
          <Button testID="btnCloud" style={styles.btnRight} title="Cloud" onPress={() => navigate('cloud')} />
        </View>
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  sectionTop: {
    borderTopWidth: 0,
  },
  sectionButtons: {
    flexDirection: 'row',
    marginBottom: Metrics.smallMargin,
  },
  btnLeft: {
    flex: 1,
    marginRight: Metrics.smallMargin,
  },
  btnRight: {
    flex: 1,
    marginLeft: Metrics.smallMargin,
  },
  textSubTitle: {
    ...Fonts.font.base.caption,
    color: Colors.transparentWhite('0.87'),
    marginBottom: 4,
  },
});

export default SelectConnect;
