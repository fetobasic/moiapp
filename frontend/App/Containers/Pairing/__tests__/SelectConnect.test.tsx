import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { mockedState } from 'App/Fixtures/mocks/mockedState';
import SelectConnect from '../SelectConnect';

const mockStore = createMockStore([]);
const store = mockStore({
  ...mockedState,
});
const props = {
  navigation: {
    navigate: jest.fn() as any,
  },
  route: { params: {} },
} as any;

describe('SelectConnect', () => {
  test('renders screen', () => {
    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <SelectConnect {...props} />
        </NavigationContainer>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders screen select btnDirect', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <NavigationContainer>
          <SelectConnect {...props} />
        </NavigationContainer>
      </Provider>,
    );

    const btnDirect = getByTestId('btnDirect');

    fireEvent.press(btnDirect);
  });
  test('renders screen select btnCloud', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <NavigationContainer>
          <SelectConnect {...props} />
        </NavigationContainer>
      </Provider>,
    );

    const btnCloud = getByTestId('btnCloud');

    fireEvent.press(btnCloud);
  });
});
