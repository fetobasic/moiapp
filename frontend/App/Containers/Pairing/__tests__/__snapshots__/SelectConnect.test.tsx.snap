// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectConnect renders screen 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "off",
      "right": "off",
      "top": "off",
    }
  }
  style={
    {
      "backgroundColor": "rgba(27, 27, 27, 1)",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(27, 27, 27, 1)",
          "borderBottomLeftRadius": 24,
          "borderBottomRightRadius": 24,
          "elevation": 12,
          "shadowColor": "rgba(0, 0, 0, 1)",
          "shadowOffset": {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 12,
          "zIndex": 1,
        },
        {
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 60,
          "justifyContent": "center",
          "marginHorizontal": 24,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={20}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            {
              "left": 0,
              "padding": 10,
              "position": "absolute",
            },
          ]
        }
      >
        <SvgMock />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginHorizontal": 48,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RobotoCondensed-Medium",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 32,
            }
          }
        >
          Select Connection
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "rgba(27, 27, 27, 1)",
        "flex": 1,
        "paddingBottom": 0,
        "paddingHorizontal": 24,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <RCTScrollView
        showsVerticalScrollIndicator={false}
      >
        <View>
          <View
            style={
              [
                {
                  "backgroundColor": "rgba(27, 27, 27, 1)",
                  "borderRadius": 16,
                  "paddingHorizontal": 12,
                  "paddingVertical": 15,
                },
                undefined,
                {
                  "marginBottom": 8,
                  "marginHorizontal": 1,
                  "paddingVertical": 16,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": 32,
                    "justifyContent": "center",
                    "width": 32,
                  }
                }
              >
                <SvgMock
                  color="rgba(255, 255, 255, 0.6)"
                />
              </View>
              <View
                style={
                  [
                    {
                      "flex": 1,
                      "paddingLeft": 12,
                      "paddingRight": 12,
                    },
                    undefined,
                    {
                      "paddingRight": 0,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "rgba(255, 255, 255, 0.87)",
                        "fontFamily": "Roboto-Medium",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                        "marginVertical": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  Cloud Connect
                </Text>
                <Text
                  style={
                    {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "Roboto-Medium",
                      "fontSize": 12,
                      "fontWeight": "300",
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                >
                  Best for home use
                </Text>
                <Text
                  style={
                    {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "Roboto-Medium",
                      "fontSize": 12,
                      "fontWeight": "300",
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                >
                  Use Cloud Connect to set up your device through your internet network. This lets you control your device from anywhere you have internet access, using your phone. It also allows feature updates.
                </Text>
                <Text
                  style={
                    {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "Roboto-Medium",
                      "fontSize": 12,
                      "fontWeight": "300",
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "rgba(255, 205, 29, 1)",
                      }
                    }
                  >
                    Note: 
                  </Text>
                  Cloud Connect requires an internet connection and Goal Zero log in.
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              [
                {
                  "backgroundColor": "rgba(27, 27, 27, 1)",
                  "borderRadius": 16,
                  "paddingHorizontal": 12,
                  "paddingVertical": 15,
                },
                undefined,
                {
                  "marginBottom": 8,
                  "marginHorizontal": 1,
                  "paddingVertical": 16,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  {
                    "borderTopWidth": 0,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": 32,
                    "justifyContent": "center",
                    "width": 32,
                  }
                }
              >
                <SvgMock
                  color="rgba(255, 255, 255, 0.6)"
                />
              </View>
              <View
                style={
                  [
                    {
                      "flex": 1,
                      "paddingLeft": 12,
                      "paddingRight": 12,
                    },
                    undefined,
                    {
                      "paddingRight": 0,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "rgba(255, 255, 255, 0.87)",
                        "fontFamily": "Roboto-Medium",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                        "marginVertical": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  Direct Connect
                </Text>
                <Text
                  style={
                    {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "Roboto-Medium",
                      "fontSize": 12,
                      "fontWeight": "300",
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                >
                  Best for off-grid use
                </Text>
                <Text
                  style={
                    {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "Roboto-Medium",
                      "fontSize": 12,
                      "fontWeight": "300",
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                >
                  Use Direct Connect to set up your device through Bluetooth. You will be able to control your device from your phone when you are within 50 feet of the device.
                </Text>
                <Text
                  style={
                    {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "Roboto-Medium",
                      "fontSize": 12,
                      "fontWeight": "300",
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "rgba(255, 205, 29, 1)",
                      }
                    }
                  >
                    Note: 
                  </Text>
                  Automatic feature updates are not available through Direct Connect.
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "marginBottom": 8,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            [
              {
                "elevation": 12,
                "shadowColor": "rgba(0, 0, 0, 1)",
                "shadowOffset": {
                  "height": 4,
                  "width": 4,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 12,
              },
              {
                "flex": 1,
                "marginRight": 8,
              },
            ],
          ]
        }
        testID="btnDirect"
      >
        <View
          style={
            {
              "borderRadius": 24,
              "overflow": "hidden",
            }
          }
        >
          <RNSVGSvgView
            bbHeight="48"
            bbWidth="100%"
            focusable={false}
            height={48}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                {
                  "flex": 0,
                  "height": 48,
                  "width": "100%",
                },
              ]
            }
            width="100%"
          >
            <RNSVGGroup
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
            >
              <RNSVGDefs>
                <RNSVGLinearGradient
                  gradient={
                    [
                      0,
                      -13224394,
                      1,
                      1776411,
                    ]
                  }
                  gradientTransform={null}
                  gradientUnits={0}
                  name="grad"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0.5"
                />
              </RNSVGDefs>
              <RNSVGRect
                fill={
                  {
                    "brushRef": "grad",
                    "type": 1,
                  }
                }
                height="48"
                propList={
                  [
                    "fill",
                  ]
                }
                width="100%"
                x="0"
                y="0"
              />
              <View
                style={
                  [
                    {
                      "borderRadius": 24,
                      "marginHorizontal": 1,
                      "marginTop": 1,
                    },
                    {
                      "backgroundColor": "rgba(27, 27, 27, 1)",
                      "height": 46,
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "paddingVertical": 12,
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "rgba(255, 255, 255, 1)",
                            "fontFamily": "Roboto-Medium",
                            "fontSize": 14,
                            "fontWeight": "700",
                            "lineHeight": 20,
                            "textTransform": "uppercase",
                          },
                          {
                            "color": "rgba(191, 210, 43, 1)",
                          },
                          undefined,
                        ]
                      }
                    >
                      Direct
                    </Text>
                  </View>
                </View>
              </View>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            [
              {
                "elevation": 12,
                "shadowColor": "rgba(0, 0, 0, 1)",
                "shadowOffset": {
                  "height": 4,
                  "width": 4,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 12,
              },
              {
                "flex": 1,
                "marginLeft": 8,
              },
            ],
          ]
        }
        testID="btnCloud"
      >
        <View
          style={
            {
              "borderRadius": 24,
              "overflow": "hidden",
            }
          }
        >
          <RNSVGSvgView
            bbHeight="48"
            bbWidth="100%"
            focusable={false}
            height={48}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                {
                  "flex": 0,
                  "height": 48,
                  "width": "100%",
                },
              ]
            }
            width="100%"
          >
            <RNSVGGroup
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
            >
              <RNSVGDefs>
                <RNSVGLinearGradient
                  gradient={
                    [
                      0,
                      -13224394,
                      1,
                      1776411,
                    ]
                  }
                  gradientTransform={null}
                  gradientUnits={0}
                  name="grad"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0.5"
                />
              </RNSVGDefs>
              <RNSVGRect
                fill={
                  {
                    "brushRef": "grad",
                    "type": 1,
                  }
                }
                height="48"
                propList={
                  [
                    "fill",
                  ]
                }
                width="100%"
                x="0"
                y="0"
              />
              <View
                style={
                  [
                    {
                      "borderRadius": 24,
                      "marginHorizontal": 1,
                      "marginTop": 1,
                    },
                    {
                      "backgroundColor": "rgba(191, 210, 43, 1)",
                      "height": 46,
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "paddingVertical": 12,
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "rgba(255, 255, 255, 1)",
                            "fontFamily": "Roboto-Medium",
                            "fontSize": 14,
                            "fontWeight": "700",
                            "lineHeight": 20,
                            "textTransform": "uppercase",
                          },
                          {
                            "color": "rgba(27, 27, 27, 1)",
                          },
                          undefined,
                        ]
                      }
                    >
                      Cloud
                    </Text>
                  </View>
                </View>
              </View>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
